tower/src/util/map_future.rs

```diff
@@ -0,0 +1,105 @@
+use std::{
+    fmt,
+    future::Future,
+    task::{Context, Poll},
+};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// [`Service`] returned by the [`map_future`] combinator.
+///
+/// [`map_future`]: crate::util::ServiceExt::map_future
+#[derive(Clone)]
+pub struct MapFuture<S, F> {
+    inner: S,
+    f: F,
+}
+
+impl<S, F> MapFuture<S, F> {
+    /// Creates a new [`MapFuture`] service.
+    pub fn new(inner: S, f: F) -> Self {
+        Self { inner, f }
+    }
+
+    /// Returns a new [`Layer`] that produces [`MapFuture`] services.
+    ///
+    /// This is a convenience function that simply calls [`MapFutureLayer::new`].
+    ///
+    /// [`Layer`]: tower_layer::Layer
+    pub fn layer(f: F) -> MapFutureLayer<F> {
+        MapFutureLayer::new(f)
+    }
+
+    /// Get a reference to the inner service
+    pub fn get_ref(&self) -> &S {
+        &self.inner
+    }
+
+    /// Get a mutable reference to the inner service
+    pub fn get_mut(&mut self) -> &mut S {
+        &mut self.inner
+    }
+
+    /// Consume `self`, returning the inner service
+    pub fn into_inner(self) -> S {
+        self.inner
+    }
+}
+
+impl<R, S, F, T, E, Fut> Service<R> for MapFuture<S, F>
+where
+    S: Service<R>,
+    F: FnMut(S::Future) -> Fut,
+    E: From<S::Error>,
+    Fut: Future<Output = Result<T, E>>,
+{
+    type Response = T;
+    type Error = E;
+    type Future = Fut;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx).map_err(From::from)
+    }
+
+    fn call(&mut self, req: R) -> Self::Future {
+        (self.f)(self.inner.call(req))
+    }
+}
+
+impl<S, F> fmt::Debug for MapFuture<S, F>
+where
+    S: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("MapFuture")
+            .field("inner", &self.inner)
+            .field("f", &format_args!("<{}>", std::any::type_name::<F>()))
```

Comment:
nice, maybe we should do this for other combinator types with function parameters! (in a separate PR).

incidentally, I tried this out in the playground and it looks like `type_name` actually returns the name of function pointers, and the location of closures, which is quite nice: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=655a2434c5ffe8a85a1126ac24d1fe9d maybe we should actually omit the `<>`s?

Reply:
Good idea. I made an issue https://github.com/tower-rs/tower/issues/545

`<>` removed in 4dc5ff0ba21fe.